tests/functional/test_redemption.py

```diff
@@ -0,0 +1,150 @@
+        redemption.redeem(
+            AMOUNT,
+            sender=panda,
+            value=estimate - estimate * SLIPPAGE_TOLERANCE // SLIPPAGE_DENOMINATOR - 1,
+        )
+    redemption.redeem(
+        AMOUNT,
+        sender=panda,
+        value=estimate - estimate * SLIPPAGE_TOLERANCE // SLIPPAGE_DENOMINATOR,
+    )
+    assert yfi.balanceOf(panda) == AMOUNT
+    with ape.reverts("price out of tolerance"):
+        redemption.redeem(
+            AMOUNT,
+            sender=panda,
+            value=estimate + estimate * SLIPPAGE_TOLERANCE // SLIPPAGE_DENOMINATOR + 1,
+        )
+    redemption.redeem(
+        AMOUNT,
+        sender=panda,
+        value=estimate + estimate * SLIPPAGE_TOLERANCE // SLIPPAGE_DENOMINATOR,
+    )
+    assert yfi.balanceOf(panda) == 2 * AMOUNT
+
+
+def test_ramp(chain, redemption, gov, panda):
+    assert redemption.scaling_factor() == AMOUNT
+    assert redemption.scaling_factor_ramp() == (0, 0, AMOUNT, AMOUNT)
+    with ape.reverts():
+        redemption.start_ramp(2 * AMOUNT, sender=panda)
+
+    ts = chain.pending_timestamp + 10
+    redemption.start_ramp(2 * AMOUNT, 1000, ts, sender=gov)
+    assert redemption.scaling_factor_ramp() == (ts, ts + 1000, AMOUNT, 2 * AMOUNT)
+    chain.pending_timestamp = ts + 200
+    chain.mine()
+    assert redemption.scaling_factor() == AMOUNT * 12 // 10
+
+    with ape.reverts():
+        redemption.start_ramp(3 * AMOUNT, sender=gov)
+
+    with ape.reverts():
+        redemption.stop_ramp(sender=panda)
+
+    chain.pending_timestamp = ts + 500
+    with chain.isolate():
+        redemption.stop_ramp(sender=gov)
+        assert redemption.scaling_factor() == AMOUNT * 15 // 10
+        assert redemption.scaling_factor_ramp() == (
+            0,
+            0,
+            AMOUNT * 15 // 10,
+            AMOUNT * 15 // 10,
+        )
+
+    chain.mine()
+    assert redemption.scaling_factor() == AMOUNT * 15 // 10
+    chain.pending_timestamp = ts + 1000
+    chain.mine()
+    assert redemption.scaling_factor() == AMOUNT * 2
+    chain.pending_timestamp = ts + 2000
+    chain.mine()
+    assert redemption.scaling_factor() == AMOUNT * 2
+
+
+def test_kill(d_yfi, yfi, redemption, gov, panda):
+    assert yfi.balanceOf(gov) == 0
+    d_yfi.mint(panda, AMOUNT * 2, sender=gov)
+    yfi.mint(redemption, AMOUNT, sender=gov)
+    redemption.kill(sender=gov)
+    assert yfi.balanceOf(gov) == AMOUNT
+    estimate = redemption.eth_required(AMOUNT)
+    d_yfi.approve(redemption, AMOUNT, sender=panda)
+    with ape.reverts("killed"):
+        redemption.redeem(AMOUNT, sender=panda, value=estimate)
+
+
+def test_sweep(d_yfi, yfi, redemption, gov):
+    d_yfi.mint(redemption, AMOUNT, sender=gov)
+    yfi.mint(redemption, AMOUNT, sender=gov)
+    redemption.sweep(d_yfi, sender=gov)
+    assert d_yfi.balanceOf(gov) == AMOUNT
+    with ape.reverts("protected token"):
+        redemption.sweep(yfi, sender=gov)
+
+
+def test_oracle(project, gov):
```

Comment:
can we add another test integrating with redemption last price check to see if the data is used correctly in integration?

say we check answers are the same as last_price reported from redemptions in the fork test.